### DISCUSSION TITLE
Threat Model for gramine project

### DISCUSSION THREAD

**Discussion by blossomin** [gramineproject/gramine#1465] [General]
Hello I want to discuss some issues about gramine's threat model:

I know that Intel SGX is for protecting workload inside the enclave from outside, such as OS, vmm. 
I'm curious about the opposite direction: can gramine libos protect itself from potential workload?
To be more exact: 
If some user loads a malicous workload in the gramine libos, can this malicous application break down the libos? 
For example, change some global states/data structures for libos internal usage.

BTW, do you know how the gramine libos isolate itself from the application running on it? I know the application can only access the gramine libos/pal interfaces by syscalls provided by libos according to the original Eurosys14 paper, but I'm not sure where this rule is enforced. I'd like to appreciate it if you could provide some suggestions.

Thanks,
Blossomin.

**Comment by dimakuv** [comment]
I'll leave very quick replies. Maybe other Gramine devs may expand on the answers.

> I'm curious about the opposite direction: can gramine libos protect itself from potential workload?

No, this is not in scope/threat model of Gramine.

> If some user loads a malicous workload in the gramine libos, can this malicous application break down the libos?

Yes, easily. E.g., there is no boundary between the Gramine code/data and the application code/data. So the app may probe its memory, find Gramine code/data and do random things with it.

> BTW, do you know how the gramine libos isolate itself from the application running on it? 

There are no isolation mechanisms.

> I know the application can only access the gramine libos/pal interfaces by syscalls provided by libos according to the original Eurosys14 paper, but I'm not sure where this rule is enforced.

I'm unsure what you mean by "enforced". Technically, nothing prevents an app to jump somewhere in the middle of Gramine code. So I guess the answer to your question is -- no, there is no enforcement of this rule.

**Reply by blossomin** [reply]
Thanks for your discussion. 

By the rules, I refer to Figure 2 in this paper(http://www.cs.unc.edu/~porter/pubs/tsai14graphene.pdf), how is this syscall restriction approach implemented?

Maybe it's too old, current gramine has removed this feature?

**Reply by dimakuv** [reply]
> By the rules, I refer to Figure 2 in this paper(http://www.cs.unc.edu/~porter/pubs/tsai14graphene.pdf), how is this syscall restriction approach implemented?

Syscall restriction is done via the seccomp filter (on `gramine-direct`; for `gramine-sgx` the Intel SGX hardware itself works as a "filter"). Gramine installs this seccomp filter into host Linux on startup. This seccomp filter basically has a rule "redirect all syscall attempts from the application back into Gramine LibOS entrypoint".

If you're interested in the exact place in the Gramine code, check here:
- https://github.com/gramineproject/gramine/blob/24319e435d6e5ec1002a5781d0f1f5362b389d32/pal/src/host/linux/pal_exception.c#L228
- https://github.com/gramineproject/gramine/blob/24319e435d6e5ec1002a5781d0f1f5362b389d32/pal/src/host/linux/pal_exception.c#L127-L185

> Maybe it's too old, current gramine has removed this feature?

No, Gramine still has this feature.

**Reply by mkow** [reply]
@blossomin: This is not a security feature, it's there only to hook `syscall` instructions.
Also, please note that this paper is almost 10 years old, most of Gramine code has been rewritten since then.

**Reply by blossomin** [reply]
Yes I do know that that paper is almoste ten years ago. But there is little material to learn the overall design for Gramine.  It will be nice if the community can update the design detail for Gramine. Thanks for your reply.

**Comment by blossomin** [comment]
Yeah, I see. The original syscall restriction works for **gramine-direct**.

To my understanding, the **gramine-sgx** defines a set of interfaces for applications inside the enclave to do some **ocall**. 

One gramine interface has two sides: one inside the enclave, and the other side outside the enclave. Can the malicous application have its own **ocall** interface and bypass the filters provided by gramine pal? For example, the attacker can provided the untrusted part of interface on the host OS, and the malicous application can have its own part of interface inside the enclave, then the application can use this its "own" interface to interact with OS.


Another confusing point about the **trusted files** in manifest: the term "trusted" means that the os can't modify or replace this file, so the application inside the enclave can trust the file content.

Potential feature for libos: current gramine libos doesn't support any isolation between workload and libos. Since Intel SGX doesn't provide another privilege level to run the libos, can we do something like software fault isolation inside the enclave to provide some "soft" privilege level and isolate the workload and libos?

Thanks your informative reply! :)

**Comment by dimakuv** [comment]
> One gramine interface has two sides: one inside the enclave, and the other side outside the enclave. Can the malicous application have its own ocall interface and bypass the filters provided by gramine pal?

Yes, the malicious app can do whatever it wants inside the SGX enclave, including having its own OCALL interface. Gramine doesn't prevent the app from doing crazy things.

> Another confusing point about the trusted files in manifest: the term "trusted" means that the os can't modify or replace this file, so the application inside the enclave can trust the file content.

Right. So what is confusing about it?

> Potential feature for libos: current gramine libos doesn't support any isolation between workload and libos. Since Intel SGX doesn't provide another privilege level to run the libos, can we do something like software fault isolation inside the enclave to provide some "soft" privilege level and isolate the workload and libos?

Yes, this is definitely possible, though Gramine doesn't provide any isolation between the app and the LibOS/PAL code. There were several projects and papers discussing such isolation mechanisms, most of them revolving around hardware-assisted SFI (Software Fault Isolation).

I remember that Occlum had an MPX-based isolation, but I don't know if Occlum still has it -- I think they abandoned the idea. Also you can google for "paper SGX SFI", I'm sure you'll find a bunch of such proposals.

Gramine is not doing this, at least currently, simply because it's not our priority and it's also quite expensive in terms of perf.

**Comment by blossomin** [comment]
Thanks for your suggestion, I will check the Occlum paper and other potential isolation methods in SGX.

**Comment by blossomin** [comment]
Sorry to bother you again. 

I want to discuss one general design about Intel SGX:

If I want to write a native enclave application, I need to specify the ecall/ocall interface in the **xxx.edl** file before we can build an enclave and launch it. 

Then my question is: after an enclave has be launched, will it be possible to construct new ecall/ocall interface? 
To my understanding, by specify these interfaces in **xxx.edl** file, the sgx sdk can generate some glue codes. If the malicious enclave application can provide/generate "ecall" interface inside the enclave and the colluded operating system will provide the glue code outside the enclave, then I guess it might be possible to have new ecall/ocall interfaces? But I'm not sure how to let the code inside the enclave know which function it wants to call outside the enclave when it wants to call ocall instruction.

Could you share some insight on this issue? 

Best,
Dong

**Reply by dimakuv** [reply]
> If the malicious enclave application can provide/generate "ecall" interface inside the enclave and the colluded operating system will provide the glue code outside the enclave, then I guess it might be possible to have new ecall/ocall interfaces?

Yes, this is generally possible.

> But I'm not sure how to let the code inside the enclave know which function it wants to call outside the enclave when it wants to call ocall instruction.

For this, you'll need to learn the OCALL/ECALL interface at the lowest level (of x86 assembly). Regarding your particular question -- when an OCALL is performed, it boils down to calling the `EEXIT` x86 instruction, and the EEXIT instruction jumps to the RIP that points to the instruction next to EENTER (when the untrusted code performed an ECALL). Moreover, on the EXIT instruction, almost all general purpose registers (GPRs) are unmodified, so you can put the "which function it wants to call" in e.g. the R10 register, and that's how the trusted enclave side will tell its intention to the untrusted side.

**Comment by mkow** [comment]
Btw.: If you want to protect your host from malicious code then just run it in a security sandbox, there are many popular solutions for that. I.e. run Gramine inside a security sandbox. The purpose of Gramine is to protect the app from the host, the other way around is out of scope for us.

**Reply by blossomin** [reply]
The key point for my case: the host and enclave code both could be malicious.  If I run a sandbox on the host, potentially malicious host could break it.

I want to find somewhere to build the trust on.

**Reply by mkow** [reply]
But isn't that still fine? The security sandbox on the host is there only to protect the host from the enclave, so the host has no interest in breaking it. Same for the other way.

**Reply by blossomin** [reply]
Sorry, my statement is not clear. Let me clarrify it.

The threat model I want to fight is: the enclave application might be malicious, and the host/OS is untrusted.
Then I want to have a SGX sandbox that can fight against the collusion of enclave application and host.

This the mutual distrust for SGX application, although the original SGX threat model considers enclave appliction to be trusted.

**Reply by mkow** [reply]
I don't think this is technically/practically feasible (even on the design/definitions level...). What are you actually trying to do? This is a very exotic threat model.

**Comment by blossomin** [comment]
We want to do something like: [this paper](https://www.usenix.org/system/files/conference/osdi16/osdi16-hunt.pdf).

Of course,we can't do this directly, but we can put some constraint on the applications.

**Reply by mkow** [reply]
I'm sorry, I don't have time to read this, but from the abstract I'm pretty sure what they are trying to achieve is impossible in practice and they have some small-print caveats somewhere.

**Reply by blossomin** [reply]
I did some surveys recently. I found some interesting research works to solve this mutual distrust scenario. I can share some overview with you:

1. SGX + MPX(Memory Protection Extension): In the original Occlum [paper](https://arxiv.org/pdf/2001.07450.pdf), they use the MPX to provide multi-domain SFI inside the enclave and support multi-process isolation in the same enclave instance. Unfortunately, MPX is deprecated now.
2. SGX + MPK(Memory Protection Key): Some researchers use MPK for memory separation inside the enclave. However, due to the incompatible trust model of MPK and SGX, where the MPK trusts OS/kernel to configure and SGX distrusts OS/kernel, there might be some collusion attack as described in Section 4.1 in [LightEnclave](https://www.usenix.org/system/files/sec22-gu-jinyu.pdf). So the LightEnclave extended the hardware and software to support intra-enclave isolation. Although their design is interesting, there is no physically available hardware.
3. SGX + SFI(Software Fault Isolation): I found several works using SFI to provide memory partition inside the enclave, like [Chancel](https://www.ndss-symposium.org/wp-content/uploads/ndss2021_7B-2_24057_paper.pdf), [Reusable Enclave for Serverless](https://www.usenix.org/conference/usenixsecurity23/presentation/zhao-shixuan). They will use compiling techniques to instrument the user application code to make sure that enclave application programs are restricted when executing. For example, memory access should be restricted to some memory areas, and the control flow's integrity should be guaranteed. I know that it's impossible to restrict the application programs if they are not instrumented or restricted. Does this SFI method make sense to you, @mkow ?

What I want to do is restrict the capability of enclave application code, like forbidding some instructions using binary scanning, and finally provide a design like the following figure, using a sandbox and SFI to confine the user application programs execution:
<img width="748" alt="截屏2023-07-27 08 37 52" src="https://github.com/gramineproject/gramine/assets/86922326/4db7ea1d-a117-48b1-acc0-585c2a39c0f7">

Do you think this is possible or not? 

And here are some properties enforce by the [Ryoan](https://dl.acm.org/doi/pdf/10.1145/3231594), I highlight the three enforced by NaCl(a kind of SFI sandbox), do these three properties make sense to you? @mkow : 

![截屏2023-07-27 13 40 58](https://github.com/gramineproject/gramine/assets/86922326/9fefe44e-6361-44e9-b71d-8c79d23a0fae)

**Reply by mkow** [reply]
> 1. SGX + MPX(Memory Protection Extension): In the original Occlum [paper](https://arxiv.org/pdf/2001.07450.pdf), they use the MPX to provide multi-domain SFI inside the enclave and support multi-process isolation in the same enclave instance. Unfortunately, MPX is deprecated now.

As you said, MPX is deprecated already.

>     2. SGX + MPK(Memory Protection Key): Some researchers use MPK for memory separation inside the enclave. However, due to the incompatible trust model of MPK and SGX, where the MPK trusts OS/kernel to configure and SGX distrusts OS/kernel, there might be some collusion attack as described in Section 4.1 in [LightEnclave](https://www.usenix.org/system/files/sec22-gu-jinyu.pdf). So the LightEnclave extended the hardware and software to support intra-enclave isolation. Although their design is interesting, there is no physically available hardware.

Yep, as you say, not possible on the current HW :)

>     3. SGX + SFI(Software Fault Isolation): I found several works using SFI to provide memory partition inside the enclave, like [Chancel](https://www.ndss-symposium.org/wp-content/uploads/ndss2021_7B-2_24057_paper.pdf), [Reusable Enclave for Serverless](https://www.usenix.org/conference/usenixsecurity23/presentation/zhao-shixuan). They will use compiling techniques to instrument the user application code to make sure that enclave application programs are restricted when executing. For example, memory access should be restricted to some memory areas, and the control flow's integrity should be guaranteed. I know that it's impossible to restrict the application programs if they are not instrumented or restricted. Does this SFI method make sense to you, @mkow ?

In a research paper? Yes, sure. But I don't believe it's viable in practice. Such protections are very brittle and hard to implement correctly, and I'm pretty sure I could escape the ones from these papers. Also, I think these papers (excluding Ryoan) don't claim to prevent covert channels between the app and the host?
And as you noted, this wouldn't work with Gramine, which was created to make running unmodified binaries easier.

> forbidding some instructions using binary scanning

This is impossible unless you restrict the instruction set to a narrow subset like Google's NaCl, but then you'd need a special compiler which emits only that verifiable x86 subset.

> and finally provide a design like the following figure, using a sandbox and SFI to confine the user application programs execution:

This design doesn't prevent covert channel communication.

> And here are some properties enforce by the [Ryoan](https://dl.acm.org/doi/pdf/10.1145/3231594), I highlight the three enforced by NaCl(a kind of SFI sandbox), do these three properties make sense to you? @mkow :

What about covert channels?
Not even discussing the complexity of the overall approach :) With something that complex you'll need years to get rid of all security vulns.
